doc.go

```diff
@@ -18,121 +18,235 @@
+//   if err := c.Provide(newDBConnection); err != nil {
+//     // ...
+//   }
 //
+// Multiple constructors can rely on the same type. Types in the container are
+// treated as singletons, so the constructors for them will be called at most
+// once when requested by another type.
 //
-// There are plans to expand the API to support multiple objects of the same type
-// in the container, but for time being consider using a factory pattern.
+//   err := c.Provide(func(conn *sql.DB) *CommentGateway {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
+//   }
 //
+// Constructors can declare any number of dependencies in their parameters and
+// optionally, return errors.
 //
-// Provide
+//   err := c.Provide(func(u *UserGateway, c *CommentGateway) (*RequestHandler, error) {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
+//   }
 //
-// The Provide method adds a constructor of an object (or objects), to the container.
-// A constructor can be a function returning any number of objects and, optionally,
-// an error.
+//   if err := c.Provide(newHTTPServer); err != nil {
+//     // ...
+//   }
 //
+// Constructors can also return multiple results to add multiple types to the
+// container.
 //
-// Each argument to the constructor is registered as a **dependency** in the graph.
+//   err := c.Provide(func(conn *sql.DB) (*UserGateway, *CommentGateway, error) {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
+//   }
 //
-//   type A struct {}
-//   type B struct {}
-//   type C struct {}
+// Invoke
 //
-//   c := dig.New()
-//   constructor := func (*A, *B) *C {
-//     // At this point, *A and *B have been resolved through the graph
-//     // and can be used to provide an object of type *C
-//     return &C{}
+// Types added to to the container may be consumed by using the Invoke method.
+// Invoke accepts a function which accepts one or more parameters and
+// optionally, returns an error. Dig calls the function with the requested
+// type, instantiating only those types that were requested by the function.
+// The call fails if any type or its dependencies (both direct and transitive)
+// were not available in the container.
+//
+//   err := c.Invoke(func(l *log.Logger) {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
 //   }
-//   err := c.Provide(constructor)
-//   // dig container is now able to provide *C to any constructor.
-//   //
-//   // However, note that in the current example *A and *B
-//   // have not been provided, meaning that the resolution of type
-//   // *C will result in an error, because types *A and *B can not
-//   // be instantiated.
 //
-// Advanced Provide
+//   err := c.Invoke(func(server *http.Server) error {
+//     // ...
+//   })
+//   if err != nil {
+//     // ...
+//   }
 //
-// // TODO: docs on dig.In usage
-// // TODO: docs on dig.Out usage
+// If the invoked function returns an error, that error is propagated to the
```

Comment:
Do we need to add a special paragraph about handling pointer and value types?
It is obvious for us, but not really for users. Although it can be handled by examples.